### ISSUE TITLE
capture by reference instead of making a copy

### ISSUE THREAD

**Issue by yiran-ever** [EOSIO/eos#5765]
@taokayan  in #5746 , you said that
> This is C/C++ macro, not a function call. Macro parameters will be replaced by the actual value before compilation.
> 
> on the other hand, `api_handle` is actually a smart pointer after pre-processing.

I'm afraid that i cannot agree with u. I don't think  `api_handle` is a pointer here. The macro can be expanded as follows, and `api_handle` is copied when the lambda-expression is evaluated. So we will create a `api_handle` object for each lambda function.
```
{
      std::string("/v1/chain/get_info"),
   
      [this, ro_api](string, string body, url_response_callback cb) mutable {
        ro_api.validate();
        try {
          if (body.empty())
            body = "{}";
          auto result = ro_api.get_info(
              fc::json::from_string(body)
                  .as<chain_apis::read_only::get_info_params>());
          cb(200l, fc::json::to_string(result));
        } catch (...) {
          http_plugin::handle_exception("chain", "get_info", body, cb);
        }
      }
}
```

**Comment by jgiszczak** [comment]
`get_read_only_api()` in `chain_plugin` constructs an instance of `chain_apis::read_only` on the stack and returns it by value.  This value is kept as a local variable in `chain_api_plugin::plugin_startup()` and will go out of scope once startup is complete.  Therefore the lambda must capture by value, or the object will be lost.  This does indeed result in independent copies of the `api_handle` object for each lambda.  This is by design.

The capabilities enabled by this design ended up not being leveraged in the current system.  Since duplicating the object is a one-time startup cost in a startup process of which it is a minor part, there's little incentive to refactor the system into a shared pointer instead.

If you wish to submit a pull request with a shared pointer implementation, we will review it.

**Comment by yiran-ever** [comment]
yea, `ro_api` and  `rw_api` are all local variables now, maybe using class members are better. I'll look into it and submit a pull request if needed.

**Comment by tbfleming** [comment]
@yiran-ever have any profiler runs or other performance measurements pointed at this code as a significant problem area?

**Comment by yiran-ever** [comment]
no, not yet.
By the way, `ro_api` & `rw_api` can be captured by reference even though they are local variables.

**Comment by tbfleming** [comment]
The lambdas outlive the function; capturing by value would create a dangling reference.

**Comment by tbfleming** [comment]
Oops typo. Capturing by reference would create a dangling reference.

**Comment by yiran-ever** [comment]
yea, i was wrong. local variables cannot be captured by reference. The lambda is not as strong as blocks in objective-c.
